File: Friends Pairing Problem.java
// Problem: Friends Pairing Problem
// geeksforgeeks problem of the day -> 7th August 2026
// JAVA CODE
class Solution {
    int[] dp;
    
    public int countFriendsPairings(int n) {
        dp = new int[n+1];
        return solve(n);
    }
    
    private int solve(int n){
        if(n <= 2){
            return n;
        }
        
        if(dp[n] != 0){
            return dp[n];
        }
        
        int single = solve(n-1);
        int pair = (n-1)*solve(n-2);
        
        return dp[n] = (single + pair);
    }
}

File: Friends Pairing Problem.cpp
// Problem: Friends Pairing Problem
// geeksforgeeks problem of the day -> 7th August 2026
// C++ CODE
class Solution {
public:
    int countFriendsPairings(int n) {
        if (n <= 2)
            return n;
        long long prev2 = 1;
        long long prev1 = 2; 
        for (int i = 3; i <= n; i++) {
            long long curr = prev1 + (i - 1) * prev2;
            prev2 = prev1;
            prev1 = curr;
        }
        return prev1;
    }
};